MapboxNavigationTests/NavigationViewControllerTests.swift

```diff
@@ -0,0 +1,192 @@
+        
+        DispatchQueue.main.asyncAfter(deadline: fireTime) {
+            
+            navigationViewController.mapView?.delegate?.mapView?(customMapView, didFinishLoading: customMapView.style!)
+            navigationViewController.mapView?.delegate?.mapViewDidFinishLoadingMap?(customMapView)
+            
+            routeController.locationManager(routeController.locationManager, didUpdateLocations: [turkStreetLocation])
+            
+            let wayNameView = (navigationViewController.mapViewController?.navigationView.wayNameView)!
+            let currentRoadName = wayNameView.text!
+            XCTAssertEqual(currentRoadName, roadName, "Expected: \(roadName); Actual: \(currentRoadName)")
+            XCTAssertTrue(wayNameView.isHidden, "WayNameView should be hidden.")
+            
+            mapviewExpectation.fulfill()
+        }
+        
+        wait(for: [mapviewExpectation], timeout: delay)
+    }
+    
+    // TODO: Most likely a UI-Test candidate.
+    func testNavigationViewControllerDelegateRoadNameAtLocationUmimplemented() {
+        
+        let dummyNavigationViewController = dependencies.dummyNavigationViewController
+        let routeController = dummyNavigationViewController.routeController!
+        
+        // Identify a location without a custom road name.
+        let fultonStreetLocation = dependencies.poi[2]
+        
+        let mapviewExpectation = expectation(description: "mapview_exp_default_roadname")
+        let customMapView = MGLMapView(frame: (dummyNavigationViewController.mapView?.bounds)!, styleURL: URL(string: "mapbox://styles/mapbox/streets-v9"))
+        
+        let delay = 1.0
+        let fireTime = DispatchTime.now() + delay
+        
+        DispatchQueue.main.asyncAfter(deadline: fireTime) {
+            
+            dummyNavigationViewController.mapView?.delegate?.mapView?(customMapView, didFinishLoading: customMapView.style!)
+            dummyNavigationViewController.mapView?.delegate?.mapViewDidFinishLoadingMap?(customMapView)
+            
+            routeController.locationManager(routeController.locationManager, didUpdateLocations: [fultonStreetLocation])
+            
+            // let roadName = "Fulton Street"
+            // let wayNameView = (dummyNavigationViewController.mapViewController?.navigationView.wayNameView)!
+            // let currentRoadName = wayNameView.text!
+            // XCTAssertEqual(currentRoadName, roadName, "Expected: \(roadName); Actual: \(currentRoadName)")
+            // XCTAssertFalse(wayNameView.isHidden, "WayNameView should be hidden.")
+            
+            mapviewExpectation.fulfill()
+        }
+        
+        wait(for: [mapviewExpectation], timeout: delay)
+    }
+}
+
+class NavigationViewControllerSpy: NavigationViewController { }
+class NavigationViewControllerFakeSpy: NavigationViewController { }
+
+extension NavigationViewControllerSpy: NavigationViewControllerDelegate {
```

Comment:
Cool, however in this case I wouldn't call this a Spy, given how we're using it. Also, it's not usually considered good form to spy on the test subject itself since we should be most concerned with how it is reacting externally to events or messages.

Could we have the test case class itself (or another test helper class) be the delegate instead?

Reply:
@akitchen Perhaps a test helper class is ideal because developers who initially extend the `NavigationViewController` class need to implement `NavigationViewControllerDelegate.navigationViewController(_:roadNameAt:)` in order to customize road names. Hence, this test ensures the validity of this new functionality.